src/server/data_source.rs

```diff
@@ -1,142 +1,197 @@
-use std::ffi::c_void;
+use std::{
+    ffi::c_void,
+    panic::{catch_unwind, AssertUnwindSafe},
+    ptr::NonNull,
+};
 
 use open62541_sys::{
     UA_Boolean, UA_DataSource, UA_DataValue, UA_NodeId, UA_NumericRange, UA_Server, UA_StatusCode,
 };
 
 use crate::{server::NodeContext, ua, DataType};
 
-/// Read callback for [`DataSource`].
-#[allow(clippy::module_name_repetitions)]
-pub type DataSourceRead = dyn FnMut(&mut DataSourceReadContext) -> DataSourceResult;
-
-/// Write callback for [`DataSource`].
 #[allow(clippy::module_name_repetitions)]
-pub type DataSourceWrite = dyn FnMut(&mut DataSourceWriteContext) -> DataSourceResult;
+pub type DataSourceResult = std::result::Result<(), ua::StatusCode>;
```

Comment:
How about introducing an `enum DataSourceError`. With a `StatusCode(ua::StatusCode)` variant and an `OperationNotSupported` variant for data sources that do not support either read or write operations.